### ISSUE TITLE
How do I use a private registry in Dockerfile FROM directive?

### ISSUE THREAD

**Issue by Infinoid** [concourse/oci-build-task#70]
I have a pipeline that works like this:
1. get app git repo (and trigger)
2. get base image (and trigger)
3. get dockerfile git repo (and trigger)
4. get oci-build image
5. run tests in base container
6. build app image from the base image using `oci-build`
7. push app image

The base image lives in a private registry, and `registry-image` is configured with `ca_certs`.  However, it seems that same image can't be used in a `Dockerfile`'s `FROM` directive.

The Dockerfile looks like this:

```dockerfile
FROM private.registry/path/thing:latest

RUN mkdir -p /app
ADD src/* /app/
# some s6 service setup stuff happens here
```

When I try to build that using `oci-build`, it fails:

```
selected worker: f2056608e065
INFO[0000] building image
#1 [internal] load build definition from Dockerfile
#1 sha256:447e72129d3e8a6381a820eaf89f26738ca3f201b845c2556d7678163aad29da
#1 transferring dockerfile: 274B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:a268ffd1544e83a8fbb87e1f4e84a378b05aa3db60ea32104495e430bdbd74df
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for private.registry/path/thing:latest
#3 sha256:01a116a8737b77c4651f411a4d20fa7da3437b841befd6c76e5b8e2098412b9a
#3 ERROR: failed to do request: Head https://private.registry/v2/path/thing/manifests/latest: x509: certificate signed by unknown authority
------
 > [internal] load metadata for private.registry/path/thing:latest:
------
error: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to do request: Head https://private.registry/v2/path/thing/manifests/latest: x509: certificate signed by unknown authority
```

If I were running docker/podman directly, I would put the cert in `/etc/docker/certs.d/`.

How do I give `oci-build-task` the CA certificate for `private.registry`?
Alternately, can the `FROM` directive refer to the image that was already fetched (using `registry-image`) earlier in the same concourse job?

**Comment by Infinoid** [comment]
> Alternately, can the `FROM` directive refer to the image that was already fetched (using `registry-image`) earlier in the same concourse job?

Following up... the above is possible.  The `$IMAGE_ARG_*` stuff in the README was a hint, but it wasn't clear how to use it.  Only after I read through [test cases](https://github.com/concourse/oci-build-task/blob/master/task_test.go#L316) and [the associated test data](https://github.com/concourse/oci-build-task/blob/master/testdata/image-args/Dockerfile) was I able to get something working.  Though I've no idea if I'm doing it in the "correct" way.

I suppose the docs can be cleaned up once the RFC is finalized.

Cool feature.  Thanks!

**Comment by sneal** [comment]
I'm using a private Harbor registry with a self signed cert. I committed the cert to my git repo and then added a custom task that wraps the oci-image task and appends the Harbor cert to the container's trust chain.

build.yml
```yaml
---
platform: linux

image_resource:
  type: registry-image
  source:
    repository: concourse/oci-build-task

inputs:
- name: s3c-repo

outputs:
- name: image

run:
  path: s3c-repo/ci/tasks/build.sh
```

build.sh
```bash
#!/usr/bin/env sh

set -eux

cp s3c-repo/ci/harbor.crt /usr/local/share/ca-certificates/harbor.crt
cat /usr/local/share/ca-certificates/harbor.crt >> /etc/ssl/certs/ca-certificates.crt

# build the container image using the oci builder task
build
```

**Comment by Infinoid** [comment]
@sneal, thanks for posting that.  I switched to building a wrapper like you have, and it simplified things quite a bit.

Seems like a good solution to the problem, closing issue.

**Comment by holgerstolzenberg** [comment]
Okay so we stubbled upon this problem also now. Trying to use a private registry. Even if the there is a workaround given, the workaround itself is neither really straight forward nor really clean in terms of usage of this resource. (No offence here ✌️)

For me the issue needs to be reopened. It should be possible to add credentials that lead to a docker login before so that private registries can be used, what is pretty much a thing in most regulated environments.

**Comment by Infinoid** [comment]
Well, client login credentials and server CA certificate are separate things, but ideally they should both be configurable.
I agree with you, the above workaround doesn't apply as nicely in the case where you need to do `docker login`.

In my mind, the best case scenario would be if `oci-build-task` takes some of the [same set of parameters `registry-image-resource` takes](https://github.com/concourse/registry-image-resource#user-content-source-configuration), specifically `username`, `password` and `ca_certs`.  `registry-image` uses those to access the remote registry.  `oci-build` would use them to access images specified in `FROM` directives.

One wrinkle is that a Dockerfile can have multiple `FROM` directives, and each one may need different credentials... I don't know the right way to deal with that.

**Comment by lannocc** [comment]
I too would love to see the same parameters (I need username and password, specifically) that @Infinoid suggested.

**Comment by taylorsilva** [comment]
What Infinoid initially mentioned is what we do in our CI. We `get` the image and then pass the image into the oci-build task using the `IMAGE_*` param.

You can see the setup in the pipeline here:
https://github.com/concourse/ci/blob/3446529e24765e6140f185f8457e3a66b0f60879/pipelines/concourse.yml#L78-L85

In our Dockerfile we start with this:
```Dockerfile
ARG base_image=concourse/golang-builder
FROM ${base_image}
```

So OP's example would look like this:
```Dockerifle
ARG base_image=private.registry/path/thing:latest
FROM ${base_image}

RUN mkdir -p /app
ADD src/* /app/
# some s6 service setup stuff happens here
```

Docs for this could be improved in the README or even on the main project website.

**Comment by lannocc** [comment]
My solution (hopefully temporary) was to copy in a pre-existing "auths" hash from a terminal where I've already done `docker login`. My task extends `oci-build-task` to save this auth in `~/.docker/config.json` and then call `build`.

I'm not sure if the auth key encodes the user/pass or if it's a session hash so I expect it may expire and this is not a long-term solution.
